Kitodo-FileManagement/src/main/java/org/kitodo/filemanagement/locking/LockManagement.java

```diff
@@ -0,0 +1,511 @@
+                throw new AccessDeniedException(permissions.getUser() + " claims to have a privilege to " + uri
+                        + " he does not have at all. Whatever he did, it was wrong.");
+            }
+            granted.checkAuthorization(permissions.getUser(), permissions.getLock(uri), write);
+        } catch (AccessDeniedException e) {
+            logger.trace("{} is not allowed to open a {} channel to {}. The reason is: {}", permissions.getUser(),
+                write ? "writing" : "reading", uri, e.getMessage());
+            throw e;
+        }
+        AbstractLock lock = permissions.getLock(uri);
+        if (lock instanceof ImmutableReadLock) {
+            uri = ((ImmutableReadLock) lock).getImmutableReadCopyURI();
+        }
+        logger.trace("{} is allowed to open a {} channel to {}.", permissions.getUser(), write ? "writing" : "reading",
+            uri);
+        return uri;
+    }
+
+    /**
+     * Closes a set of permissions. First, all locks are checked for their
+     * release. If that fails, an IOException is thrown. Otherwise, all locks
+     * are removed from the permission set and possible temporary files are
+     * logged off. This process can be parallelized.
+     *
+     * @param user
+     *            user who returns the locks
+     *
+     * @param locks
+     *            the locks that are returned
+     * @throws IllegalStateException
+     *             if there is still a stream flowing over the lock
+     */
+    void close(String user, Map<URI, AbstractLock> locks) {
+        synchronized (this) {
+            canClose(locks.keySet());
+            locks.entrySet().parallelStream().map(entry -> entry.getKey())
+                    .peek(uri -> removePermission(user, locks.get(uri), uri))
+                    .forEach(uri -> immutableReadFileManagement.maybeRemoveReadFile(uri, user));
+        }
+    }
+
+    /**
+     * Closes a permission from a set.
+     *
+     * @param access
+     *            record from which the authorization is to be removed
+     * @param uri
+     *            URI for which the authorization is to be closed
+     * @throws IllegalStateException
+     *             if there is still a stream flowing over the lock
+     */
+    void close(GrantedAccess access, URI uri) {
+        synchronized (this) {
+            canClose(Arrays.asList(uri));
+            removePermission(access.getUser(), access.getLock(uri), uri);
+            immutableReadFileManagement.maybeRemoveReadFile(uri, access.getUser());
+            access.forgetAccessTo(uri);
+        }
+    }
+
+    /**
+     * Captures a flow of pairs into a map.
+     *
+     * @param <K>
+     *            type of map key
+     * @param <V>
+     *            type of map value
+     * @param pairStream
+     *            stream of pairs (future map entries)
+     * @return a map containing the pairs
+     */
+    private <K, V> Map<K, V> collectPairsToMap(Stream<Pair<K, V>> pairStream) {
```

Comment:
Class usage would save you such an operation.

Reply:
It seems to me, own class for this is a bit of overkill, but if you want ...

Reply:
It seems to me, own class for this is a bit of overkill, but if you want ...